service/history/historyReplicator.go

```diff
@@ -68,167 +77,238 @@ func newHistoryReplicator(shard ShardContext, historyEngine *historyEngineImpl,
+		return nil
+	}
+	// currentLastWriteVersion <= incomingVersion
+	logger.Debugf("Retrying replication task. Current RunID: %v, Current LastWriteVersion: %v, Incoming Version: %v.",
+		currentRunID, currentLastWriteVersion, incomingVersion)
+	return ErrRetryEntityNotExists
+}
 
-		// Check for duplicate processing of replication task
-		if firstEvent.GetEventId() < msBuilder.GetNextEventID() {
-			replicationState := msBuilder.GetReplicationState()
-			r.metricsClient.IncCounter(metrics.ReplicateHistoryEventsScope, metrics.StaleReplicationEventsCounter)
-			logger.Debugf("Dropping replication task.  State: {NextEvent: %v, Version: %v, LastWriteV: %v, LastWriteEvent: %v}",
-				msBuilder.GetNextEventID(), replicationState.CurrentVersion,
-				replicationState.LastWriteVersion, replicationState.LastWriteEventID)
-			return nil
+func (r *historyReplicator) ApplyOtherEventsVersionChecking(context *workflowExecutionContext, msBuilder mutableState,
+	request *h.ReplicateEventsRequest, logger bark.Logger) (mutableState, error) {
+	var err error
+	// check if to buffer / drop / conflict resolution
+	incomingVersion := request.GetVersion()
+	replicationInfo := request.ReplicationInfo
+	rState := msBuilder.GetReplicationState()
+	if rState.LastWriteVersion > incomingVersion {
+		// Replication state is already on a higher version, we can drop this event
+		// TODO: We need to replay external events like signal to the new version
+		logger.Warnf("Dropping stale replication task. CurrentV: %v, LastWriteV: %v, LastWriteEvent: %v, IncomingV: %v.",
+			rState.CurrentVersion, rState.LastWriteVersion, rState.LastWriteEventID, incomingVersion)
+		return nil, nil
+	} else if rState.LastWriteVersion < incomingVersion {
+		// Check if this is the first event after failover
+		logger.Infof("First Event after replication. CurrentV: %v, LastWriteV: %v, LastWriteEvent: %v, IncomingV: %v.",
+			rState.CurrentVersion, rState.LastWriteVersion, rState.LastWriteEventID, incomingVersion)
+		previousActiveCluster := r.clusterMetadata.ClusterNameForFailoverVersion(rState.LastWriteVersion)
+		ri, ok := replicationInfo[previousActiveCluster]
+		if !ok {
+			logger.Errorf("No ReplicationInfo Found For Previous Active Cluster. Previous Active Cluster: %v, Request Source Cluster: %v, Request ReplicationInfo: %v.",
+				previousActiveCluster, request.GetSourceCluster(), request.ReplicationInfo)
+			// TODO: Handle missing replication information, #840
+			return nil, ErrMissingReplicationInfo
 		}
 
-		// Check for out of order replication task and store it in the buffer
-		if firstEvent.GetEventId() > msBuilder.GetNextEventID() {
-			logger.Debugf("Buffer out of order replication task.  NextEvent: %v, FirstEvent: %v",
-				msBuilder.GetNextEventID(), firstEvent.GetEventId())
-			r.metricsClient.IncCounter(metrics.ReplicateHistoryEventsScope, metrics.BufferedReplicationTaskCounter)
+		// Detect conflict
+		if ri.GetLastEventId() < rState.LastWriteEventID {
```

Comment:
you changed '!=' to '<'.  Whats the reasoning?  Can we capture that as a comment.

Reply:
done